Add Tachyon support for package managed installations

smoogipoo opened the pull request:
Resolves https://github.com/ppy/osu/issues/33565
Supersedes / closes https://github.com/ppy/osu/pull/33645

- Considers up to 10 items from the first page. This is also what [`velopack`](https://github.com/velopack/velopack/blob/10c80a835e8b8ceef42f546b3ed5506d35d7428d/src/lib-csharp/Sources/GithubSource.cs#L98-L109) does.
- Adds `OSU_EXTERNAL_UPDATE_STREAM`. Packaged builds will use this to set their release flavour, for example [`osu-lazer-tachyon-bin`](https://aur.archlinux.org/packages/osu-lazer-tachyon-bin) would set it to `Tachyon`, and [`osu-lazer-bin`](https://aur.archlinux.org/packages/osu-lazer-bin) would set it to `Lazer`.
- Basically rewrote `UpdateManager` to better handle notifications and cancellations and all that.

smoogipoo committed:
Move release stream handling to base class

smoogipoo committed:
Adjust packaged update manager to check release stream

smoogipoo committed:
Add `OSU_EXTERNAL_UPDATE_STREAM`

Valid values are `Tachyon`/`Lazer`.

peppy commented on osu.Game/Updater/NoActionUpdateManager.cs:
```diff
@@ -30,11 +32,17 @@ protected override async Task<bool> PerformUpdateCheck()
+                ReleaseStream stream = externalReleaseStream ?? ReleaseStream.Value;
+                bool includePrerelease = stream == Configuration.ReleaseStream.Tachyon;
+
+                OsuJsonWebRequest<GitHubRelease[]> releasesRequest = new OsuJsonWebRequest<GitHubRelease[]>("https://api.github.com/repos/ppy/osu/releases?per_page=10&page=1");
```
I can imagine scenarios where 5 may be too few (if we push too many consecutive prerelease builds) so it's something we need to be very wary off. Maybe the answer is to have one prerelease live at max, then limit this to 2..

peppy commented on osu.Game/Updater/UpdateManager.cs:
```diff
@@ -62,11 +66,14 @@ protected override void LoadComplete()
 
-        private readonly object updateTaskLock = new object();
+            config.BindWith(OsuSetting.ReleaseStream, releaseStream);
+            releaseStream.BindValueChanged(_ => scheduleUpdateCheck());
```
Hmm, this does have a slight edge case which is that if there's a check already running, it will assume that another doesn't need to be run (even thought it probably does because the release stream changed). Might be worth fixing?

See logic in `CheckForUpdateAsync` below when `updateCheckTask` is not null and gets used again as `waitTask`.

smoogipoo replied:
Check if you agree with the new logic. I had to adjust the inner workings of `UpdateManager` a little bit.

smoogipoo committed:
Adjust `UpdateManager` to keep a check stream

There are two primary paths here:
1. A user requests an update. The existing request is used if it's in
progress, or a new request is made and processed immediately.
2. Something is changed (e.g. the release stream) that triggers a
background request. A new request is made to run after the existing one.

smoogipoo committed:
Add tests

smoogipoo committed:
Partial classes

smoogipoo committed:
Expand tests to perform a second check

To ensure it doesn't get stuck in a state where new checks aren't
performed.

smoogipoo committed:
Make check task non-nullable

Keeping unnecessary nullables around is a pet-peeve of mine. This
simplifies things.

smoogipoo committed:
Rewrite `UpdateManager` to handle cancellations

smoogipoo committed:
Allow cancelling from initial notification

smoogipoo committed:
Merge branch 'master' into package-manager-tachyon

smoogipoo committed:
Remove no-longer existing argument

I'm not entirely sure how this works, but CI was testing against the
updated Velopack package, whereas my local package was outdated and had
4 args.

Previous commit merges master to update the package version, this commit
fixes the args.

smoogipoo commented:
Oooooookay. It's going to be tough to review but I've basically had to completely rewrite `UpdateManager` to fix the above issues. At the end of the day it seems the issue is that it fundamentally needs cancellation support, so I've added that now.

As a result, I've made this supersede and close https://github.com/ppy/osu/pull/33645.

Please consider this effectively a new PR at this point.

---
Here's a demonstration of cancellations:

https://github.com/user-attachments/assets/b4d2cd87-8fac-4dd0-be89-211b69910534

1. Shows startup update check still working
2. Cancel update during download
    - **Now supported**.
    - On `master` this has no effect.
3. Cancel from initial notification
    - **Now supported**.
    - On `master` this has no effect.
4. Click "check for updates" during update
    - **Now supported**: cancels in-progress update and checks again.
    - On `master` the button is disabled during the update.
5. Cancel update during download
    - Same as 2.
6. Click "check for updates" after download completes
    - **Now supported**: nullifies existing update, closes active notification, checks again. 
    - On `master` the notification remains alive.

Not shown above is proving that the scheduled update still works:

https://github.com/user-attachments/assets/a915255e-9e70-4a86-b5c4-f97951ffedf2

Not shown above is proving that changing release streams works. You're just going to have to trust me on this, but it will act the same as clicking the "check for updates" button.

smoogipoo committed:
Resolve warnings in tests

smoogipoo commented on osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:
```diff
@@ -143,8 +143,9 @@ private async Task checkForUpdates()
             {
-                // This sequence allows the notification to be immediately dismissed.
-                checkingNotification.State = ProgressNotificationState.Cancelled;
+                // This sequence allows the notification to be immediately dismissed without posting a continuation message.
+                checkingNotification.CompletionTarget = null;
+                checkingNotification.State = ProgressNotificationState.Completed;
```
This change has significance because setting `Cancelled` also cancels the internal CTS that then cancels the entire rest of the update process in crossfire.

`Completed` doesn't do this but instead posts a completion message. This is resolved by setting `CompletionTarget`.

smoogipoo commented on osu.Desktop/Updater/VelopackUpdateManager.cs:
```diff
@@ -27,143 +27,117 @@ public partial class VelopackUpdateManager : Game.Updater.UpdateManager
         private bool isInGameplay => localUserInfo?.PlayingState.Value != LocalUserPlayingState.NotPlaying;
 
-        private readonly Bindable<ReleaseStream> releaseStream = new Bindable<ReleaseStream>();
-        private UpdateManager? updateManager;
-        private UpdateInfo? pendingUpdate;
```
One of the main points is that this entire class has become stateless with regards to the update.

On `master`, the statefulness means that if an update is found for the "Lazer" stream this class would _refuse_ to check for a subsequent update on the "Tachyon" stream.

smoogipoo committed:
Dispose the last CTS

smoogipoo committed:
Don't check for updates in DEBUG

smoogipoo commented on osu.Game/Updater/UpdateManager.cs:
```diff
@@ -62,35 +66,44 @@ protected override void LoadComplete()
+        {
+            var cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            var lastCancellation = Interlocked.Exchange(ref updateCancellation, cancellation);
 
-            lock (updateTaskLock)
-                updateCheckTask = null;
+            using (lastCancellation)
+            {
+                // This serves a dual purpose of nullifying the last update, closing any existing notifications as stale.
+                await lastCancellation.CancelAsync().ConfigureAwait(false);
+            }
```
I battled internal demons here, and I'm still not sure of the best way to do this.

One alternative I considered is is asserting `ThreadSafety.IsUpdateThread` and doing away with the thread-safety. I quickly discarded that when I made the mistake of `Task.Run(CheckForUpdateAsync)` in the scheduled background check. 
That particular code instead uses the synchronous method above, but it's still something I consider easy to get wrong, so I settled on this somewhat convoluted way of handling thread safety. Hopefully it's not too hard to look at.

smoogipoo replied:
I now consider the above comment moot, because I discovered the same being done by SSV2:

https://github.com/ppy/osu/blob/d03b5b7b0b196bce8f89435bff89017d7e974dc9/osu.Game/Graphics/Carousel/Carousel.cs#L319-L322

The difference here is `using (lastCancellation)`, which is required because this is a linked CTS.

peppy committed:
Update cancellation token naming / inline comment slightly

peppy commented on osu.Desktop/Updater/VelopackUpdateManager.cs:
```diff
@@ -27,143 +27,117 @@ public partial class VelopackUpdateManager : Game.Updater.UpdateManager
-                    scheduleRecheck = true;
-                    notification.FailDownload();
-                    Logger.Error(e, @"update failed!");
+                    await updateManager.DownloadUpdatesAsync(update, p => progressNotification.Progress = p / 100f, cts.Token).ConfigureAwait(false);
```
i do wonder if we want this to also happen outside of gameplay..

peppy committed:
Add prefix to log events
